src/cyclebane/graph.py

```diff
@@ -24,18 +24,21 @@ def _get_new_node_name(graph: nx.DiGraph) -> str:
         if name not in graph:
             return name
 
-
 def _remove_ancestors(graph: nx.DiGraph, node: Hashable) -> nx.DiGraph:
-    graph = graph.copy()
+    graph_without_node = graph.copy()
+    graph_without_node.remove_node(node)
     ancestors = nx.ancestors(graph, node)
-    ancestors_successors = {
-        ancestor: graph.successors(ancestor) for ancestor in ancestors
+    # Considering the graph we obtain by removing `node`, we need to consider the
+    # descendants of each ancestor. If an ancestor has descendants that are not
+    # removal candidates, we should not remove the ancestor.
+    ancestors_descendants = {
+        ancestor: nx.descendants(graph_without_node, ancestor) for ancestor in ancestors
     }
```

Comment:
You could use 
```suggestion
    ancestors_descendants = [
        (ancestor, nx.descendants(graph_without_node, ancestor)) for ancestor in ancestors
    ]
```
to avoid computing hashes. And technically, you could also express the whole construction of `to_remove` as a list comprehension. (I'm surprised ruff didn't suggest the latter.)